src/lang/nodejs/src/index.js

```diff
@@ -0,0 +1,199 @@
+ */
+function setPort(customPort) {
+  validifyPort(customPort);
+  port = customPort;
+  return {
+    visualise: visualise,
+    openManually: openManually
+  };
+}
+
+/**
+ * Stops the browser from automatically opening
+ */
+function openManually() {
+  openBrowser = false;
+  return {
+    visualise: visualise,
+    setPort: setPort
+  };
+}
+
+/**
+ * Renders a file as part of a response to a request
+ * @param {*} response Object representing the response
+ * @param {String} filepath Path to the file to be rendered
+ * @param {String} contentType The media type of the file
+ */
+function renderFile(response, filepath, contentType) {
+  fs.readFile(filepath, (err, data) => {
+    if (err) {
+      console.log(err);
+      response.writeHead(500, {
+        'Content-Type': 'text/plain'
+      });
+      response.end('500 - Internal Server Error');
+    } else {
+      response.writeHead(200, {
+        'Content-Type': contentType
+      });
+      response.end(data, 'utf-8');
+    }
+  });
+}
+
+/**
+ * Creates and sets up a server which listens on the specified port
+ * @param {Number} port Port on which the server listens
+ */
+function startServer(port) {
+  httpServer = http.createServer((req, res) => {
+    switch (req.url) {
+      case '/':
+        renderFile(res, HTML_FILE_PATH, 'text/html');
+        break;
+      case '/' + JS_FILE_PATH:
+        renderFile(res, JS_FILE_PATH, 'text/javascript');
+        break;
+      default:
+        res.writeHead(404, {
+          'Content-Type': 'text/plain'
+        });
+        res.end('404 - Page not Found');
+        break;
+    }
+  });
+
+  httpServer.listen(port, (req, res) => {
+    console.log('Server listening on port ' + port);
+
+    // show index.html in the browser
+    if (openBrowser) {
+      open(HOST + ':' + port);
+    }
+  });
+}
+
+/**
+ * Sets up the websocket on the server end. Defines event handlers for web socket connection.
+ * @param {*} jsonStr json data passed as argument to first visualise call
+ */
+function setWebsocket() {
+  const webSocket = new WebSocketServer({
+    httpServer: httpServer
+  });
+  webSocket.on('request', (request) => {
+    connection = request.accept(null, request.origin);
+
+    connection.on('close', () => console.log('Connection closed'));
+
+    connection.on('message', (message) => console.log(message));
+
+    waitingData.forEach((data) => {
+      connection.send(formatJSON(data));
+    });
+  });
+}
+
+/**
+ * Validates that the argument passed to visualise is a valid JSON string
+ * @param {*} jsonStr Argument passed by user to visualise
+ */
+function validifyJSON(jsonStr) {
+  // CHECK IF STRING IS JSON
+  try {
+    JSON.parse(jsonStr);
+  } catch (e) {
+    const type = typeof jsonStr;
+    if (type !== 'object' && type !== 'string') {
+      throw new Error('Visualise must take in a valid JSON value');
```

Comment:
Let's define the error messages in another js file and import them. So that they can be used by the tests as well without duplication/the maintenance of having to keep them identical.

Reply:
but the tests are in a separate file

Reply:
Wait so you want to export the errors? that sounds a bit weird